apisix/discovery/zookeeper/init.lua

```diff
@@ -0,0 +1,238 @@
+           log.error("failed to create parent: ", cerr)
+        end
+    elseif err then
+        utils.close_zk_client(client)
+        log.error("exist zk root failed: ", err)
+        return
+    end
+
+    -- 3.  All instance nodes under a service
+    local services, err = client:get_children(zookeeper_conf.root_path)
+    if not services then
+        utils.close_zk_client(client)
+        log.error("get zk root children failed: ", err)
+        return
+    end
+
+    -- 4. Fetch the instances of each service and update the cache
+    local now = ngx.time()
+    for _, service in ipairs(services) do
+        local instances, err = fetch_service_instances(zookeeper_conf, service)
+        if instances then
+            instance_cache:set(service, {
+                nodes = instances,
+            }, now + zookeeper_conf.cache_ttl)
+        else
+            log.error("fetch service instances failed: ", service, " err: ", err)
+        end
+    end
+
+    -- 5. Close connects
+    utils.close_zk_client(client)
+end
+
+function _M.nodes(service_name)
+    if not service_name or service_name == "" then
+        log.error("service name is empty")
+        return nil
+    end
+
+    -- 1. Check instance_cache initialized
+    if not instance_cache then
+        log.error("instance_cache not initialized")
+        return nil
+    end
+
+    -- 2. Get from cache
+    local cache, _, _ = instance_cache:get(service_name)
+    local now = ngx.time()
+
+    -- 3. If the cache is missed or expired, actively pull (the data)）
+    if not cache or cache.expire_time < now then
+        log.debug("cache miss or expired, fetch from zk: ", service_name)
+        local instances, err = fetch_service_instances(zookeeper_conf, service_name)
+        if not instances then
+            log.error("fetch instances failed: ", service_name, " err: ", err)
+            -- Fallback: Return the old cache (if available)）
+            if cache then
+                return cache.nodes
+            end
+            return nil
+        end
+
+        -- Update the cache
+        cache = {
+            nodes = instances,
+        }
+
+        instance_cache:set(service_name, cache, now + zookeeper_conf.cache_ttl)
+    end
+
+    return cache.nodes
+end
+
+function _M.init_worker()
+    -- Load configuration
+    zookeeper_conf = local_conf.discovery and local_conf.discovery.zookeeper or {}
+
+    -- The default values
+    zookeeper_conf.connect_string = zookeeper_conf.connect_string or "127.0.0.1:2181"
+    zookeeper_conf.fetch_interval = zookeeper_conf.fetch_interval or 10
+    zookeeper_conf.session_timeout = zookeeper_conf.session_timeout or 30000
+    zookeeper_conf.connect_timeout = zookeeper_conf.connect_timeout or 5000
+    zookeeper_conf.cache_ttl = zookeeper_conf.cache_ttl or 30
+    zookeeper_conf.root_path = zookeeper_conf.root_path or "/apisix/discovery/zk"
+    zookeeper_conf.auth = zookeeper_conf.auth or {type = "digest", creds = "",}
```

Comment:
We can handle default values ​​in jsonschema; you can refer to the schemas of other service discovery components.